parsers/BD.py

```diff
@@ -1,299 +1,320 @@
+            parser="BD.py",
+            message=("Could not find table body in returned HTML."),
+        )
+    row_data = parse_table_body(table_body)
 
-    XLS_SPAN = (datetime(2018, 1, 12), datetime(2018, 1, 13), datetime(2018, 1, 14))
+    return row_data
 
-    XLS_END = False
-    if shifted_target_datetime.naive in XLS_SPAN:
-        XLS_END = True
 
-    UNDERSCORE = False
-    if shifted_target_datetime <= arrow.get("20150703", "YYYYMMDD"):
-        UNDERSCORE = True
+@refetch_frequency(timedelta(days=1))
+def fetch_production(
+    zone_key: str = "BD",
+    session: Session = Session(),
+    target_datetime: Optional[datetime] = None,
+    logger: Logger = getLogger(__name__),
+) -> Union[Dict[str, Any], List[Dict[str, Any]]]:
 
-    day = shifted_target_datetime.format("DD")
-    month = shifted_target_datetime.format("MM")
-    short_year = shifted_target_datetime.format("YY")
+    row_data = query(session, target_datetime, logger)
 
-    # NOTE %20 padders needed, format is day-month-year
-    if OLD_FORMAT and UNDERSCORE:
-        URL = "https://pgcb.org.bd/PGCB/upload/Reports/Daily_Report{}-{}-{}.xls".format(
-            day, month, short_year
-        )
-    elif OLD_FORMAT:
-        URL = (
-            "https://pgcb.org.bd/PGCB/upload/Reports/Daily%20Report{}-{}-{}.xls".format(
-                day, month, short_year
-            )
-        )
-    elif XLS_END:
-        URL = "https://pgcb.org.bd/PGCB/upload/Reports/Daily%20Report%20{}-{}-{}.xls".format(
-            day, month, short_year
-        )
-    else:
-        URL = "https://pgcb.org.bd/PGCB/upload/Reports/Daily%20Report%20{}-{}-{}.xlsm".format(
-            day, month, short_year
-        )
+    production_data_list = []
+    for row in row_data:
 
-    if OLD_FORMAT:
-        df = pd.read_excel(URL, sheet_name="En.Curve", skiprows=[0, 1, 2, 3])
-        df = old_format_converter(df)
-    else:
-        if XLS_END:
-            df = pd.read_excel(
-                URL,
-                sheet_name="YesterdayGen",
-                skiprows=[0, 1, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12],
-            )
-            df = new_format_converter(df, logger)
-        else:
-            df = pd.read_excel(URL, sheet_name="YesterdayGen", skiprows=[0, 1, 3])
-            df = new_format_converter(df, logger)
+        # Create data with empty production
+        data = {
+            "zoneKey": zone_key,
+            "datetime": row["time"],
+            "production": {},
+            "source": "erp.pgcb.gov.bd",
+        }
 
-    return df, OLD_FORMAT
+        # And add sources if they are present in the table
+        known_sources_sum_mw = 0.0
+        for source_type in ["coal", "gas", "hydro", "oil", "solar"]:
+            if row[source_type] is not None:
+                # also accumulate the sources to infer 'unknown'
+                known_sources_sum_mw += row[source_type]
+                data["production"][source_type] = row[source_type]
+
+        # infer 'unknown'
```

Comment:
I just noticed that 'unknown' in the ElectricityMaps data is actually equal to the sum of both exchanges from India. I checked multiple datapoints in the graph.

Don't have time to create an issue right now, but perhaps someone else can.

Reply:
I'll try and take a quick look at this on Monday and see if it's an issue in the parser.